### PR TITLE
Add a mutex to make the fight detector threadsafe

### PR DESCRIPTION
**Pull request by nan-yu**

### CODE REVIEW

File: pkg/syncer/reconcile/fight_detector.go [hunk 1]
```diff
@@ -57,12 +58,15 @@ func FightWarning(frequency float64, resource client.Object) status.ResourceErro
 // Instantiate with newFightDetector().
 //
 // Performance characteristics:
-//  1. Current implementation is NOT threadsafe.
+//  1. Current implementation is threadsafe.
 //  2. Current implementation has unbounded memory usage on the order of the
 //     number of objects the Syncer updates through its lifetime.
 //  3. Updating an already-tracked resource requires no memory allocations and
 //     take approximately 30ns, ignoring logging time.
 type fightDetector struct {
```
Comment (sdowell) [inline pkg/syncer/reconcile/fight_detector.go:66]: I think perhaps this should be defined in it's own package with public methods to safeguard against future changes directly accessing the protected variables. Currently it looks like this is all used within the same package, so the private fields aren't really protected semantically.

Comment (nan-yu) [reply pkg/syncer/reconcile/fight_detector.go]: That'a s great point. Done.

File: pkg/syncer/reconcile/fight/logger.go [hunk 2]
```diff
@@ -21,28 +21,28 @@ import (
...
+// Logger is used to log errors about fights from Detector at most
 // once every 60 seconds. It has similar performance characteristics as
-// fightDetector.
+// Detector.
 //
-// Instantiate with newFightLogger().
-type fightLogger struct {
-	// lastLogged is when fightLogger last logged about a given API resource.
+// Instantiate with NewLogger().
+type Logger struct {
```
Comment (sdowell) [inline pkg/syncer/reconcile/fight/logger.go:29]: Should this be private? I assume it's only intended to be used within this package

Comment (nan-yu) [reply pkg/syncer/reconcile/fight/logger.go:29]: Well, it is instantiated in `pkg/syncer/reconcile/apply.go`. Maybe I should just move the instantiation into the detector.

Comment (nan-yu) [reply pkg/syncer/reconcile/fight/logger.go]: Done

**Pull request merged**